Freezing Torchscript modules

bzinodev opened the pull request:
This patch enables folding GetAttr nodes with their corresponding
values. _jit_pass_freeze_module API returns a new TorchScipt module
where all function calls and get attributes are inlined.
Usage:

frozen_model = torch._C._freeze_module(scrited_model._c)
frozen_model.forward(...)

This API currently optimizes the forward method. We will follow up to
to preserve and optimize methods and attributes that are annotated as
 @torch.jit.interface.

Several future improvements to JIT optimizations are required to maximize
clean up/de-sugar the graph and eliminate redundancies.
Ideally, we want to produce a graph that can easily be lowered to
GLOW and other low-level backends.
__

Commit:
Freezing Torchscript modules

This patch enables folding GetAttr nodes with their corresponding
values. _jit_pass_freeze_module API returns a new TorchScipt module
where all function calls and get attributes are inlined.
Usage:

frozen_model = torch._C._freeze_module(scrited_model._c)
frozen_model.forward(...)

This API currently optimizes the forward method. We will follow up to
to preserve and optimize methods and attributes on demand.

Several future improvements to JIT optimizations are required to maximize
clean up/de-sugar the graph and eliminate redundancies.
this is an important step toward producing a graph that can easily be lowered to
GLOW and other low-level backends.
__

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit ff249479:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ff249479075948004b0ed5fbfc14484c48d07858).

### :detective: 3 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4240459) pytorch_parallelnative_linux_xenial_py3_6_gcc5_4_build (1/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=85960284) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=85960284))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/docker/build.sh 
Auto-merging .circleci/docker/build.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/pytorch_build_definitions.py 
Auto-merging .circleci/cimodel/data/pytorch_build_definitions.py 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/pytorch_build_data.py 
Auto-merging .circleci/cimodel/data/pytorch_build_data.py 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/caffe2_build_definitions.py 
Auto-merging .circleci/cimodel/data/caffe2_build_definitions.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4240455) pytorch_linux_xenial_py3_6_gcc5_4_build (2/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=85960285) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=85960285))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/docker/build.sh 
Auto-merging .circleci/docker/build.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/pytorch_build_definitions.py 
Auto-merging .circleci/cimodel/data/pytorch_build_definitions.py 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/pytorch_build_data.py 
Auto-merging .circleci/cimodel/data/pytorch_build_data.py 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/caffe2_build_definitions.py 
Auto-merging .circleci/cimodel/data/caffe2_build_definitions.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4240445) pytorch_paralleltbb_linux_xenial_py3_6_gcc5_4_build (3/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=85960286) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=85960286))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/docker/build.sh 
Auto-merging .circleci/docker/build.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/pytorch_build_definitions.py 
Auto-merging .circleci/cimodel/data/pytorch_build_definitions.py 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/pytorch_build_data.py 
Auto-merging .circleci/cimodel/data/pytorch_build_data.py 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/caffe2_build_definitions.py 
Auto-merging .circleci/cimodel/data/caffe2_build_definitions.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 commented on test/test_jit.py:
```diff
@@ -1564,6 +1564,178 @@ def forward(self, x):
                    .check('GetAttr[name="_quantized_weight"]') \
                    .run(m._c._get_method('forward').graph)
 
+    @_tmp_donotuse_dont_inline_everything
```
nit: we can remove this macro now, we don't inline by default

jerryzh168 commented on torch/csrc/jit/init.cpp:
```diff
@@ -194,6 +195,11 @@ void initJITBindings(PyObject* module) {
+          [](script::Module& module) {
+            return freezeModule(module);
+          },
+          py::arg("module"))
```
nit: `.def("_jit_pass_freeze_module", &freezeModule)`
also `freezeModule` --> `FreezeModule` for consistent naming?

bzinodev replied:
freeze_module is not a pass. Once we are confident with freeze_module. There should be a follow up patch to  expose freeze_module API as public.

I agree I will  make it consistent  :)

jerryzh168 commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,305 @@
+  // are not folded and 'training' attribute is also not folded.
+  // TODO: Determine if freezing in training s useful and clarify its semantics.
+  TORCH_INTERNAL_ASSERT(!module.is_training());
+  auto moduleClone = module.clone();
```
talking about clone, does `FreezeModule` work for shared ClassTypes?

bzinodev replied:
Yes, please see the freezing tests that I have .Are  you thinking of some example?

jerryzh168 commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,305 @@
+    for (auto& name : attrsToRemove) {
+      module_._ivalue()->unsafeRemoveAttr(name);
+      module_.type()->unsafeRemoveAttribute(name);
+    }
```
I think you probably need to extend this for shared ClassType, similar to https://github.com/pytorch/pytorch/pull/31408

bzinodev replied:
I changed one of my testcases to verify current implementation works fine. The current patch cleans up only top level module. submodules that contains mutable attributes are  not clean up. 

I plan  to implement a follow up patch to do that and that case, shared ClassType need to be carefully handled.

eellison reviewed (commented):
Only took a brief look at alias analysis changes. As far as I understand, the semantics of freezing are:

"I will only run this frozen method after freezing" - which means that we must consider mutable attributes within the method, but any attributes which are not mutated we can inline. 

If the pass correctly takes account of attributes which aren't mutated and then inlines them, why are there being changes made to alias analysis?

Edit: this still doesn't handle interface calls. I think we should just throw if we see one for now, since interfaces still aren't part of the public api.

bzinodev commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -337,6 +338,8 @@ void AliasDb::analyzeImpl(Node* node) {
     case prim::GetAttr:
     case prim::unchecked_cast:
+      if (isFrozen_ && node->kind() == prim::GetAttr)
+        return analyzeCreator(node);
```
The alias analysis is conservative and treat attributes as mutable (wildcard).  Freezing allows us to assume that they cannot mutate from outside. basically I am telling the alias analysis I am loading a constant (i.e. pretending that I replaced GetAttr with prim::Constant). Then the job of alias analysis to check if the constant is mutated (inplace).

eellison replied:
I think that will fail for something like:
```
class Mod(nn.Module):
     self.my_val = [1, 2, 3]
def forward():
     val = self.my_val
     val.append(4)
     val2 = self.my_val
     val2.append(5)
     return val
```     

Each call of `self.my_val` will create a new alias, and `val2` will appear to be unused and get DCE'd/

bzinodev replied:
done

suo reviewed (changes requested):
Looking pretty close. I left some commentary inline. We should also do a quick style review after this; but I don't want to clutter up the discussion on the actual logic.

suo commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,307 @@
+          return false;
+        }
+      }
+      attr = std::move(lst);
```
we need to make sure this logic is sufficient, e.g. does it cover tuples that may contain tensors?

bzinodev replied:
Done

suo commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,307 @@
+  TORCH_INTERNAL_ASSERT(!module.is_training());
+  auto moduleClone = module.clone();
+  AttributePropagator attrPropagator(moduleClone);
+  script::Method method = moduleClone.get_method("forward");
```
Noting from offline discussion: if we are exposing more than just `forward()` of the frozen module, we need to do interprocedural analysis to make sure that some method (e.g. `mutate_foo()`) is not mutating an attr that got folded by another (`use_foo()`).

bzinodev replied:
Correct! the interface is a follow up patch :)

suo commented on test/jit/test_module_interface.py:
```diff
@@ -405,3 +405,46 @@ def forward(self, input):
+
+    # The call to forward of proxy_mod cannot be inlined. Making sure
+    # Freezing is perserving proxy_mod.
+    def test_freeze_module_with_interface(self):
```
note from offline discussion: we are not always able to inline interface calls; we should throw for now if we can't

bzinodev replied:
Throwing an error as we discussed

eellison commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -51,7 +51,8 @@ bool AliasDb::isContainerType(const TypePtr& type) {
 
-AliasDb::AliasDb(std::shared_ptr<Graph> graph) : graph_(std::move(graph)) {
+AliasDb::AliasDb(std::shared_ptr<Graph> graph, bool isFrozen)
+    : graph_(std::move(graph)), isFrozen_(isFrozen) {
```
I don't think `prim::FrozenGetAttr` is right either.

suo replied:
Oh yeah, we need to work out what happens when two attributes (`self.a` and `self.b`) alias. The current patch does not handle this case and potentially produces incorrect results by falsely inlining.

bzinodev replied:
Added a bunch of tests. Also Filed a bug because JIT is NOT preserving aliasing in __init (e.g. self.a = self.b)! 
Thanks Elias,  All your conditions are applied through preserveAttrs_  hashing.

eellison replied:
It's not preserving aliasing in lists, but it does with Tensors.

bzinodev replied:
Correct! this will be a follow up patch.

eellison reviewed (commented):
I think in order to inline attributes of a module the following needs to happen.

For a module value `%ModVal` of type `Mod` with a Tensor field `weights`:
   - no value with `Mod` type has any `prim::SetAttr[field="weights"]` (*1)
   - all `prim::GetAttr["weights"](%ModVal)` can be resolved statically
	- For example, if `%ModVal` is an output of a control flow node we cannot resolve all GetAttr 

   - `weights` is not aliased by any other Tensor contained within the frozen module, including in Lists and Tuples of Tensors (*2)
   	- You can check this by checking the Tensor storage of each tensor

   -  at this point, you can replace all `prim::GetAttr[field="weights"]` with a single top-level `prim::FrozenGetAttr`. This preserves aliasing. If the vallue doesn't have any writers it can be inlined as a constant.

(*1) You can relax this condition for a specific value of type `MyMod` if you can prove it doesn't alias any `MyMod` values that reassign `weights`, maybe better as a follow

(*2) You can relax this condition if you can prove the other aliases aren't mutated, also maybe better as a follow up. 

Edit: the above correctness condition ensures that there is one single alias for an attribute Tensor, meaning that it will work even if the Tensor is mutated. 

A separate correctness condition is to check that all aliases of an attribute Tensor are not mutated

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,307 @@
+  // corresponding value. Based on initial test on resnet50 and other torch
+  // vision tests. GetAttrs are not too frequent so it is ok to chase GetAttr
+  // chain to retrieve their values.
+  bool findConstantAttr(
```
As we discussed. the reason they kept together is because  it allows to shortcut as soon as it find out one of the submodule is mutable.

iseeyuan commented on test/test_jit.py:
```diff
@@ -1564,6 +1564,187 @@ def forward(self, x):
                    .check('GetAttr[name="_quantized_weight"]') \
                    .run(m._c._get_method('forward').graph)
 
+    def test_fold_quantize_freeze(self):
```
Is there a test of save/load of the freezed model? Something like:
```
ref = m.forward(input)
mf = torch._C._freeze_module(m._c)
mf.save(buf) # save frozen model
mf_loaded = load(buf) # load frozen model
output = mf_loaded.forward(input) # forward on loaded model
assertEQ(ref, output)
```

bzinodev replied:
Added

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,307 @@
+    }
+  }
+
+  std::set<std::string> recordReferencedAttrs(std::shared_ptr<Graph>& graph) {
```
We do need it because we need to preserve the top level sub modules that contains mutable attributes. In the first recording step, we record only leaf attributes (for example  in example below only attribute"b" as mutable). In the second step, we record attributes that contain mutable attributes (such as "sub" attribute)  to preserve them.  

class Submodule (...)
      def __init_
           self.a = 
           self.b = 
      def forward(self, ...):
           self.b +=1
           return self.a

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,307 @@
+
+  // If Module is in eval mode, detach and override requires_grad tensor
+  // attributes to enable freezing.
+  bool shouldFoldAttr(IValue& attr, std::string& name, bool isEval) {
```
Agreed. I simplified  it and generalized the function to handle tuples and lists.

eellison commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -502,6 +502,10 @@ struct CAFFE2_API IValue final {
   TypePtr type() const;
 
+  size_t hash() const {
+    return payload.as_int;
```
Why are we doing this ?

bzinodev replied:
To track aliased attributes. For now the JIT is not preserving aliasing (except for tensors) but in future the current logic can deal with: 
self.a = [1,2]
self.b = self.a

eellison commented on test/jit/test_freezing.py:
```diff
@@ -0,0 +1,421 @@
+        m_s = torch.jit.script(m)
+        m_s.eval()
+        m_f = torch._C._freeze_module(m_s._c)
+        # if should be assertTrue. scripting is not preserving aliasing (see #33034)
```
Scripting does preserve aliasing wrt Tensors.

bzinodev replied:
Fixed

eellison reviewed (commented):
Just did a quick skim. While you are right that when we script a module we construct a new list when we convert a python list to an ivalue. However that doesn't mean that each list is a unique alias.
```
class MyMod(nn.Module):
     def __init__(self):
          self.x = [1, 2, 3]
          self.y = [1]

     def make_alias(self):
          self.x = self.y
...
mod = torch.jit.script(MyMod())
mod.make_alias()
freeze(mod)
```

Agreed this is a pretty unusual case. however the more important thing here is that aliasing of Tensors is preserved. So two lists can contain the same tensor.
   
I think we need to work on / finish up the hashing of potentially aliasing ivalues still.

ZolotukhinM commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,337 @@
+  // following:
+  // 1) Remove unused attributes.
+  // 2) Remove unreferenced submodules
+  // 3) Remove non pulic unreferenced methods.
```
Typo: public.

Commit:
Clean up frozen TorchScript Modules

This patch adds two APIs to delete methods from modules and module's types.
NOTE these APIs is only for internal only uses and should be used only by freezing
where a new module and type are being created.

eellison commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -504,12 +504,45 @@ struct CAFFE2_API IValue final {
+    }
+  };
+
+  using HashAliasedIValues = std::unordered_set<IValue, HashIValue, CompIValues>;
```
Likewise, would prefer if this was moved into freezing.

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,395 @@
+    }
+    // FIXME: Current Alias analysis fails to track subvalues.
+    // This is not a common scenario, for freezing, detect and error out.
+    for (auto it = usedAttrs.begin(); it != usedAttrs.end();) {
```
This is going to be quite slow because it's O(n^2) in # of usedAttrs, and `overlap` is O(n) in number of elements.
We could instantiate a new HashAliasedIValues, and then one by one iterate over usedAttrs, inserting all of it's subvalues into the new HashedAliasIValuess, checking if the element was already contained ? This will be O(n) instead of O(n^2).

This isn't strictly necessary, if there aren't many attributes this won't be that slow, but if we are recursing through other modules than it might become an issue.

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,395 @@
+          auto inputNode = n->inputs()[0]->node();
+          auto name = n->s(attr::name);
+          auto attrModule = module_;
+          if (!findConstantAttr(inputNode, name, attrModule)) {
```
It's a little confusing that we're calling `findConstantAttr`, in the function that sets up the correctness conditions for `findConstantAttr`. Maybe add comment. Are we doing this to not re-traverse mutable attributes ?

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,395 @@
+            auto type = n->output()->type();
+            // Do not record submodules. Their attributes are tracked
+            // individually.
+            if (attr.isObject() || !AliasDb::mutableType(attr.type())) {
```
I don't know why we're skipping submodules here.

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,395 @@
+          TORCH_INTERNAL_ASSERT(attrModule.hasattr(name));
+          Value* paramConst = nullptr;
+          auto I = attrValues.find(attrModule._ivalue());
+          if (I != attrValues.end()) {
```
nit: the `I` / `II` variable names are confusing to read.

bzinodev replied:
Agreed, sorry in LLVM we capitalize variables. I am renaming I to iter and II to iter2

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,395 @@
+  script::Module& module_;
+
+  // Contains the attributes names (e.g. {"self", "subModule", "a"}
+  std::deque<std::string> names_;
```
Not sure why this is a class attribute, probably shouldn't be

bzinodev replied:
I used it to pretty print (e.g.  %self.sub.conv = prim::constant(...). the time of computing the name and inserting the new node are far apart.

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,395 @@
+    }
+  }
+
+  IValue overrideGradient(IValue attr) {
```
Can we not just call `getSubValues()`, and iterate through tensor subvalues, in-place removing gradient ?

bzinodev replied:
Yes cleaner and it handles missing cases.

bzinodev replied:
Eliminate tensor detach?

eellison commented on test/jit/test_freezing.py:
```diff
@@ -0,0 +1,602 @@
+                self.c = (self.a, 10)
+
+            def forward(self, x):
+                self.b[1] += 10
```
nice

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,395 @@
+  void run(std::shared_ptr<Graph>& graph) {
+    Inline(*graph);
+    propagateAttributes(graph);
+    runOptimization(graph, /* unroll? */ false);
```
I know quantization and potentially other parties wanted to run optimizations immediately after freezing. It might be worth removing the runOptimization call here to make freezing more composable.

bzinodev replied:
The optimization is important for the clean up. Getting rid of all unused attributes. I think it is better to return a clean graph and user can call whatever followup otimization post freezing.

eellison reviewed (approved):
🚢🚢🚢🚢🚢🚢🚢🚢🚢

Looks great, let's ship this! I left some comments that I would like addressed / responded to before landing, but let's merge this.  It covers all of the reasonable edge cases, and as you have run into, requires some structural changes to Alias Analysis to handle the remaining degenerate cases that shouldn't block the PR, + we safely error out in those cases anyway. 

Maybe you can help me here, but let's at least establish what some of the follows up:

- maybe clean up / move public apis on ivalue into freezing
- once alias analysis has fine-grained containment tracking, we can improve freezing here instead of erroring out
- remove methods, attributes

anything else ?

bzinodev commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -65,6 +65,60 @@ TypePtr IValue::type() const {
+      break;
+    case Tag::Future:
+    case Tag::Device:
+    case Tag::Object:
```
correct! I wanted to be conservative. At this point freezing does not need this case. I will add in the future with a nice testcase.

Commit:
t

eellison commented on test/jit/test_freezing.py:
```diff
@@ -0,0 +1,602 @@
+        v.append(4)
+        m_s.a = v
+        m_s.eval()
+        m_f = torch._C._freeze_module(m_s._c)
```
Also, before this pass is made public we need a workflow that does not involve accessing private members (`torch._C` and `m_s._c`). This pass should be in `torch.jit.script` with proper documentation (e.g. https://github.com/pytorch/pytorch/blob/master/torch/jit/__init__.py#L1117).

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -348,6 +349,8 @@ void AliasDb::analyzeImpl(Node* node) {
     case prim::ListUnpack:
     case prim::PythonOp:
     case prim::GetAttr:
+      if (isFrozen_ && node->kind() == prim::GetAttr)
```
This is a correctness issue and should be fixed before this is made public

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -0,0 +1,395 @@
+  // folded.
+  // TODO: Determine if freezing in training mode is useful and further clarify
+  // its semantics.
+  TORCH_CHECK(!module.is_training());
```
Could we improve the error message here ?

bzinodev commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -504,12 +504,45 @@ struct CAFFE2_API IValue final {
   TypePtr type() const;
 
+  size_t hash() const {
+    return payload.as_int;
```
I am renaming the HashIValue to HashAliasedIValue and CompIValue to CompAliasedIValue. Hope this clear enough :)
